sys/kern/kern_proc.c

```diff
@@ -2554,6 +2554,145 @@ sysctl_kern_proc_c18n(SYSCTL_HANDLER_ARGS)
 	PRELE(p);
 	return (error);
 }
+
+/*
+ * The implementation of proc_read_string() above does not stop at null
+ * terminators, which we would like to do. Return -1 on failure (e.g., fault)
+ * and otherwise the number of bytes read and a properly terminated (albeit
+ * possible truncated) string.
+ */
+static int
+proc_read_string_properly(struct thread *td, struct proc *p,
+    const char * __capability sptr, char *buf, size_t len)
+{
+	ssize_t readlen;
+	size_t n;
+
+	KASSERT(len >= 1, ("%s: Buffer too short", __func__));
+	if (len < 1)
+		return (-1);
+	for (n = 0; n < len - 1; n++) {
+		if (!cheri_can_access(sptr, CHERI_PERM_LOAD,
+		    (__cheri_addr ptraddr_t)&sptr[n], 1))
+			return (-1);
+		readlen = proc_readmem(td, p,
+		    (__cheri_addr vm_offset_t)&sptr[n], &buf[n], 1);
+		if (readlen != 1)
+			return (-1);
+		if (buf[n] == '\0')
+			break;
+	}
+	/* Unconditionally enforce termination. */
+	buf[n++] = '\0';
+	return (n);
+}
+
+/*
+ * If usefully accessible, return a c18n compartment list from the target
+ * process.
+ */
+static int
+sysctl_kern_proc_c18n_compartments(SYSCTL_HANDLER_ARGS)
+{
+	int error, *name = (int *)arg1;
+	u_int namelen = arg2;
+	struct proc *p;
+	struct cheri_c18n_info info;
+	struct cheri_c18n_compart compart;
+	char * __capability namep;
+	char * __capability namepp;
+	size_t len, i, gen;
+
+	if (namelen != 1)
+		return (EINVAL);
+
+	error = pget((pid_t)name[0], PGET_WANTREAD, &p);
+	if (error != 0)
+		return (error);
+
+	if ((p->p_flag & P_SYSTEM) != 0 ||
+	    SV_PROC_FLAG(p, SV_CHERI) == 0 ||
+	    p->p_c18n_info == NULL)
+		goto out;
+
+	len = proc_readmem_cap(curthread, p, (vm_offset_t)p->p_c18n_info, &info,
+	    sizeof(info));
+	/*
+	 * If there is a version mismatch or the compartment array is malformed,
+	 * error out.
+	 */
+	if (len != sizeof(info) ||
+	    info.version != CHERI_C18N_INFO_VERSION ||
+	    info.comparts_gen % 2 != 0 ||
+	    info.comparts_entry_size < sizeof(namep)) {
+		error = ENOEXEC;
+		goto out;
+	}
+
```

Comment:
You really want to optimize the first call to sysctl() with a NULL oldptr from userspace that is just trying to query the size to use.  You can do a decent estimate by just reading in the top-level structure with the count of compartments and using that to compute the size needed and use that with SYSCTL_OUT with a NULL input buffer and then return early without reading the actual compartment list from userspace.